### PR TITLE
CP-17268 integrate coverage profiling

### PR DESCRIPTION
**Pull request by lindig**
This change adds optional coverage profiling using `bisect_ppx`. The code changes don't affect the normal (non-profiling) case. To set up the build environment for a profiling build, do:

    make coverage
    make

The first step rewrites the `_tags` file to enable code instrumentation. It can be reverted with

    make uncover

See the COVERAGE.md for general information about coverage profiling. Supporting code lives in `profiling/` -- it doesn't have an effect when not profiling as it only sets up an environment variable. The build requires `bisect_ppx` to be available. This has been added to the SPEC file and `oasis` as well (although this is not used).

An important code change is a handler for the `SIGTERM` signal to make sure the profiling code writes it data to a file before the daemon terminates.

### CODE REVIEW

File: libvirt/xenops_libvirt_main.ml [hunk 1]
```diff
@@ -16,6 +16,7 @@ let specific_essential_paths = Path.hvm_guests @ Path.network_configuration
 
 (* Start the program with the libvirt backend *)
 let _ =
+  Coverage.init "xenopsd-libvirt";
```
Comment (euanh) [inline libvirt/xenops_libvirt_main.ml]: Indentation does not seem right here.

File: xc/xenops_xc_main.ml [hunk 2]
```diff
@@ -46,6 +46,7 @@ let make_var_run_xen () =
 
 (* Start the program with the xen backend *)
 let _ =
+  Coverage.init "xenopsd-xc"; (* set up coverage profiling *)
```
Comment (euanh) [inline xc/xenops_xc_main.ml]: Indentation.

File: xl/xenops_xl_main.ml [hunk 3]
```diff
@@ -50,6 +50,7 @@ let make_var_run_xen () =
 
 (* Start the program with the xenlight backend *)
 let _ =
+  Coverage.init "xenopsd-xl";
```
Comment (euanh) [inline xl/xenops_xl_main.ml]: Indentation.

**Comment by euanh** [comment]
Since you are touching all the oasis auto-generated files, this would be a good opportunity to switch to `oasis setup -setup-update dynamic` and remove most of the auto-generated stuff.

File: _oasis [hunk 4]
```diff
@@ -29,8 +29,53 @@ Library xenopsd
...
+    xcp.storage,
+    sexplib,
+    sexplib.syntax,
+    xcp,
+    uutf,
+    xenstore,
+    xenstore.unix,
+    xenstore_transport,
+    xenstore_transport.unix,
+    oclock
```
Comment (euanh) [inline _oasis]: I'm glad you have split these out onto separate lines!   Much better for resolving merges. :)

File: README.md [hunk 5]
```diff
@@ -9,3 +9,22 @@ xenopsd manages VMs running
...
 and provides a simple RPC control interface to the layer above (typically xapi).
 
+## Coverage Profiling
+
+This branch adds coverage profiling with bisect_ppx. Binaries write
+`bisect*.out` files, when they exit, to the current directory. These files
+can be analysed with bisect-ppx-report:
+
+    bisect-ppx-report -I _build -html coverage     *.out
+    bisect-ppx-report -I _build -text coverage.txt *.out
```
Comment (euanh) [inline README.md]: Could you add a Makefile target for this?

File: lib/xenopsd.ml [hunk 6]
```diff
@@ -137,9 +123,25 @@ let configure ?(specific_options=[]) ?(specific_essential_paths=[]) ?(specific_n
...
+      ; "/var/tmp"
+      ] in
+    let is_dir  = function 
+    | ""    -> false
+    | path  -> try Sys.is_directory path with Sys_error _ -> false
+    in try
+      List.find is_dir dirs
+    with
+      Not_found -> error "can't find temp directory %s" __LOC__; exit 1
+  in try
```
Comment (euanh) [inline lib/xenopsd.ml]: Why not use `Filename.get_temp_dir_name`?

File: lib/xenopsd.ml [hunk 7]
```diff
@@ -137,9 +123,25 @@ let configure ?(specific_options=[]) ?(specific_essential_paths=[]) ?(specific_n
...
+    | ""    -> false
+    | path  -> try Sys.is_directory path with Sys_error _ -> false
+    in try
+      List.find is_dir dirs
+    with
+      Not_found -> error "can't find temp directory %s" __LOC__; exit 1
+  in try 
     ignore (Sys.getenv "BISECT_FILE") 
   with Not_found ->
     Unix.putenv "BISECT_FILE" (tmpdir // Printf.sprintf "bisect-%s" name)
```
Comment (euanh) [inline lib/xenopsd.ml]: Can't you replace this with something like:
```
try 
   ignore (Sys.getenv "BISECT_FILE")
with Not_found ->
   Unix.putenv "BISECT_FILE" (Filename.temp_file (Printf.sprintf "bisect-%s" name) "out")
```

At least, you can replace the `let tmpdir = ...` with Filename.get_temp_dir_name`.

File: _oasis (hunk 4 above)
Comment (lindig) [inline _oasis]: I'd like to move to `oasis setup -setup-update dynamic` in a separate step. It is not in all projects straight forward but I have tested that the proposed method works in that context, too. It's done here: https://github.com/lindig/luna-rossa.

Comment (euanh) [inline _oasis]: That's fine.   Let's make sure we come back to it before we wrap this work up.

**Comment by simonjbeaumont** [comment]
@lindig if you think you've addressed all the review comments so far then please remove the "needs updating" label and ping @euanh

**Comment by lindig** [comment]
Can we get this merged? Since this is a testbed for more repos I'd like to see this exercised in a test.

File: _tags [hunk 8]
```diff
@@ -1,9 +1,3 @@
-# START_COVERAGE
-# coverage analysis with bisect_ppx
-# compile and link with package bisect_ppx
-<**/*.ml{,i,y}>: pkg_bisect_ppx
-<**/*.native>:   pkg_bisect_ppx
-# END_COVERAGE
```
Comment (simonjbeaumont) [inline _tags]: This is what I'm most nervous about with this method.

**Comment by simonjbeaumont** [comment]
If @euanh is happy then I'm happy.

**Comment by simonjbeaumont** [comment]
https://github.com/xapi-project/xenopsd/pull/267/commits/49f10245ffacdbfb442a3d35d3ef05a251f702f3 This makes me nervous at how easy it is to check this file in accidentally

**Comment by lindig** [comment]
I agree. That's why it would be better in the long run to build regular and instrumented binaries in the same build. As we discussed, I'd suggest we look into the way `ocamlbuild` infers compilation for top-lvel `xenopsd-xc.coverage` binary that gets build along with `xenopsd-xc.native`.

**Comment by jonludlam** [comment]
There are still a couple of indentation slips that @euanh noticed that still need fixing. Other than that, I agree that switching modes is dangerous, but this approach is at least nicer than some we've had in the past. If we can get ocamlbuild to do a parallel coverage-enabled build all in one go, that's definitely the way we should be going, but that might take a bit more time than we have right now to sort out.

**Pull request merged**